var/spack/repos/builtin/packages/edm4hep/package.py

```diff
@@ -61,12 +61,20 @@ class Edm4hep(CMakePackage):
         description="Use the specified C++ standard when building.",
     )
 
+    variant(
+        "json",
+        default=True,
+        description="Build edm4hep with JSON support and edm4hep2json",
+        when="@0.99.2:",
+    )
+
     depends_on("cmake@3.3:", type="build")
     depends_on("cmake@3.23:", type="build", when="@0.10.3:")
     depends_on("python", type="build")
 
     depends_on("root@6.08:")
-    depends_on("nlohmann-json@3.10.5:")
+    depends_on("nlohmann-json@3.10.5:", when="@0.99.2: +json")
+    depends_on("nlohmann-json@3.10.5", when="@:0.99.1")
```

Comment:
1. Fix typo: The version range still starts with `@3.10.5:` (not only `3.10.5`):
2. As `nlohmann-json`, a header-only library, you could also use `type="build"` for it.

```suggestion
    with default_args(type="build"):
        depends_on("nlohmann-json@3.10.5:", when="@0.99.2: +json")
        depends_on("nlohmann-json@3.10.5:", when="@:0.99.1")
```

Reply:
Thanks for the catch, that should definitely be an open range. What does the `with default_args(type="build")` do in this case?

Reply:
@tmadlener It means that 
```py
with default_args(type="build"):
    depends_on("nlohmann-json@3.10.5:", when="@0.99.2: +json")
    depends_on("nlohmann-json@3.10.5:", when="@:0.99.1")
```
is the same as:
```py
depends_on("nlohmann-json@3.10.5:", type="build", when="@0.99.2: +json")
depends_on("nlohmann-json@3.10.5:", type="build", when="@:0.99.1")
```
By default, at dependency is not just a build dependency, but also a runtime dependency.

The effects of adding `type="build"` are twofold
- At runtime, `"nlohmann-json@3.10.5:"` does not need to be installed.
- Also, when other packages depend on `edm4hep` they do not automatically get "nlohmann-json@3.10.5:" (and as the same version of `edm4hep` was built with), but either no "nlohmann-json" (if they don't depend on it), or in theory, they could get to use a different version of "nlohmann-json" (if the build was done separately, guess). 

This would be no problem because "nlohmann-json" only consists of C++ header files and not of a shared or static library. Hence, this header-only library would be needed onl when compiling, and as the headers become part of the compilation itself, there is no runtime dependency.

But you can just ignore this part of my suggestion if you are not sure about it.

Reply:
> Thanks for the catch, that should definitely be an open range. What does the with default_args(type="build") do in this case?

As per your decision that you don't want `nlohmann-json@<your-version-range>` to be a build-only dependency, but to be provided to your dependent packages as well, updated this suggestion to not use `type="build"`.

Now, it only fixed the version to not be pinned but an open range, like you said in your initial answer in this conversation.

I think you want and should apply this suggestion in this updated form.